File: python/multicast/cliente_multicast.py
import socket
import struct

OFFSET = 54  # MESMO valor do servidor (secao 3.3)

GRUPO_MULTICAST = "230.0.0.1"
# Base 4646 (nao 4446): a porta 4500 e reservada pelo Windows (IPsec NAT-T) e descarta o multicast.
PORTA = 4646 + OFFSET  # 4700


def ip_local():
    # Descobre o IP da interface de rede local (nao envia nada; so resolve a rota de saida).
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    try:
        s.connect(("8.8.8.8", 80))
        return s.getsockname()[0]
    finally:
        s.close()


IP = ip_local()

sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP)
sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
sock.bind(("", PORTA))

grupo = socket.inet_aton(GRUPO_MULTICAST)
# Inscreve-se no grupo pela interface de rede local (o mesmo IP que o servidor usa para enviar).
interface = socket.inet_aton(IP)
solicitacao_membro = struct.pack("4s4s", grupo, interface)
sock.setsockopt(socket.IPPROTO_IP, socket.IP_ADD_MEMBERSHIP, solicitacao_membro)

print(f"[Multicast] Inscrito no grupo {GRUPO_MULTICAST}:{PORTA}. Aguardando avisos...")
while True:
    dados, endereco = sock.recvfrom(1024)
    print(f"[Multicast] Recebido: {dados.decode('utf-8')}")

File: java/multicast/ClienteMulticast.java
import java.net.*;
import java.io.IOException;

public class ClienteMulticast {
    static final int OFFSET = 54; // MESMO valor do servidor (secao 3.3)

    public static void main(String[] args) throws IOException {
        String grupoMulticast = "230.0.0.1";
        // Base 4646 (nao 4446): 4500 e reservada pelo Windows (IPsec NAT-T). 4646 + OFFSET(54) = 4700.
        int porta = 4646 + OFFSET; // 4700

        try (MulticastSocket socket = new MulticastSocket(porta)) {
            InetAddress grupo = InetAddress.getByName(grupoMulticast);
            InetSocketAddress endpointGrupo = new InetSocketAddress(grupo, porta);

            // Inscreve-se no grupo pela interface de rede local. Se estiver em Wi-Fi corporativa/VPN
            // que bloqueia multicast, veja a secao 6.5 do roteiro (troubleshooting).
            NetworkInterface interfaceRede = NetworkInterface.getByInetAddress(InetAddress.getLocalHost());

            socket.joinGroup(endpointGrupo, interfaceRede);
            System.out.println("[Multicast] Inscrito no grupo " + grupoMulticast + ":" + porta + ". Aguardando avisos...");

            byte[] buffer = new byte[1024];
            while (true) {
                DatagramPacket pacote = new DatagramPacket(buffer, buffer.length);
                socket.receive(pacote);
                String mensagem = new String(pacote.getData(), 0, pacote.getLength());
                System.out.println("[Multicast] Recebido: " + mensagem);
            }
        }
    }
}
